src/align.jl

```diff
@@ -120,103 +120,193 @@ function collect_leaflet_residues(chain, tms, extracellular)
-    P = OffsetMatrix{NWParent}(undef, 0:m, 0:n)
-    # Initialize the scoring. We need to use all rows of D while always incrementing the columns.
-    # Give maximum badness to paths that don't do this.
-    S[:,0] .= typemax(T); S[0,:] .= zero(T)
-    P[:,0] .= NONE; P[0,:] .= NONE
-    for i = 1:m
-        for j = 1:i-1
-            S[i,j], P[i,j] = typemax(T), NONE
+    T = typeof(zero(eltype(D)) + zero(eltype(gapcosts)))
+    S = OffsetMatrix{T}(undef, 0:m, 0:n)            # S[i, j] is the optimal score of align(seq1[1:i], seq2[1:j])
+    P = OffsetMatrix{NWParent}(undef, 0:m, 0:n)     # parent (predecessor of position i,j)
+    fill!(S, -1)
+    # Note: UP implies we're advancing sequence 1, and thus sequence 2 is in a gap
+    #       LEFT is the converse
+    # Initialize the scoring
+    S[0,0] = zero(T); P[0,0] = NONE
+    S[1,0] = saturating_add2(gapcosts.open2, gapcosts.extend2); P[1:m,0] .= UP
+    for i = 2:m
+        S[i,0] = saturating_add2(S[i-1,0], gapcosts.extend2)
+    end
+    S[0,1] = saturating_add2(gapcosts.open1, gapcosts.extend1); P[0,1:n] .= LEFT
+    for j = 2:n
+        S[0,j] = saturating_add2(S[0,j-1], gapcosts.extend1)
+    end
+    # Compute the score to each midpoint position
+    for i = 1:m, j = 1:n
+        left = saturating_add2(S[i, j-1], gapcosts.extend1)
+        if P[i, j-1] != LEFT
+            left = saturating_add2(left, gapcosts.open1)
         end
-        for j = n-m+i+1:n
-            S[i,j], P[i,j] = typemax(T), NONE
+        up = saturating_add2(S[i-1,j], gapcosts.extend2)
+        if P[i-1, j] != UP
+            up = saturating_add2(up, gapcosts.open2)
         end
+        diag = saturating_add2(S[i-1,j-1], D[i, j])
+        minscore = min(diag, left, up)
+        P[i, j] = minscore == diag ? DIAG :
+                  minscore == left ? LEFT : UP
+        S[i, j] = minscore
     end
-    # Compute the score of each possible path
-    for i = 1:m, j = i:n-m+i
-        d = D[i,j]
-        sd, sl = S[i-1,j-1], S[i, j-1]
-        if sd == typemax(T)
-            S[i,j], P[i,j] = sl, LEFT
+    return S, P
+end
+
+function traceback_nw(P; start=last(CartesianIndices(P)))
+    # Trace the path
+    ϕ = Tuple{Int,Int}[]
+    i, j = Tuple(start)
+    istop, jstop = first.(axes(P))
+    while i > istop && j > jstop
+        p = P[i, j]
+        if p == DIAG
+            push!(ϕ, (i, j))
+            i -= 1
+            j -= 1
+        elseif p == LEFT
+            j -= 1
+        elseif p == UP
+            i -= 1
         else
-            diag = sd + d
-            left = sl
-            S[i,j], P[i,j] = diag < left ? (diag, DIAG) : (left, LEFT)
+            break
         end
     end
-    return S, P
+    return reverse!(ϕ)
+end
+
+# Saturating arithmetic to prevent overflow of Integer types
+# https://discourse.julialang.org/t/the-performance-of-saturating-operations-or-adding-intrinsics/48575
+# (thanks!)
+saturating_add2(x::T, y::T) where {T <: Unsigned} = x + min(~x, y)
```

Comment:
Thanks for including the link! This is something I often forget about, since we often work with floats